RecoCTPPS/ProtonReconstruction/src/ProtonReconstructionAlgorithm.cc

```diff
@@ -0,0 +1,388 @@
+/****************************************************************************
+ * Authors:
+ *   Jan Kašpar
+ *   Laurent Forthomme
+ ****************************************************************************/
+
+#include "RecoCTPPS/ProtonReconstruction/interface/ProtonReconstructionAlgorithm.h"
+
+#include "FWCore/MessageLogger/interface/MessageLogger.h"
+
+#include "DataFormats/CTPPSDetId/interface/CTPPSDetId.h"
+#include "DataFormats/ProtonReco/interface/ProtonTrackExtra.h"
+
+#include "TF1.h"
+
+using namespace std;
+using namespace edm;
+
+//----------------------------------------------------------------------------------------------------
+
+ProtonReconstructionAlgorithm::ProtonReconstructionAlgorithm(bool fit_vtx_y, unsigned int verbosity) :
+  verbosity_(verbosity),
+  fitVtxY_(fit_vtx_y),
+  initialized_(false),
+  fitter_(new ROOT::Fit::Fitter), chiSquareCalculator_(new ChiSquareCalculator)
+{
+  // initialise fitter
+  double pStart[] = { 0, 0, 0, 0 };
+  fitter_->SetFCN(4, *chiSquareCalculator_, pStart, 0, true);
+}
+
+//----------------------------------------------------------------------------------------------------
+
+void ProtonReconstructionAlgorithm::init(const std::unordered_map<unsigned int, LHCOpticalFunctionsSet> &opticalFunctions)
+{
+  // reset cache
+  release();
+
+  // prepare helper objects
+  unique_ptr<TF1> ff(new TF1("ff", "[0] + [1]*x"));
+
+  // build optics data for each object
+  for (const auto &p : opticalFunctions)
+  {
+    const LHCOpticalFunctionsSet &ofs = p.second;
+
+    // make record
+    RPOpticsData rpod;
+    rpod.optics = &p.second;
+    rpod.s_xi_vs_x_d = make_shared<TSpline3>("",
+      (double *) ofs.getFcnValues()[LHCOpticalFunctionsSet::exd].data(),
+      (double *) ofs.getXiValues().data(), ofs.getXiValues().size());
+    rpod.s_y_d_vs_xi = ofs.getSplines()[LHCOpticalFunctionsSet::eyd];
+    rpod.s_v_y_vs_xi = ofs.getSplines()[LHCOpticalFunctionsSet::evy];
+    rpod.s_L_y_vs_xi = ofs.getSplines()[LHCOpticalFunctionsSet::eLy];
+
+    // calculate auxiliary data
+    LHCOpticalFunctionsSet::Kinematics k_in = { 0., 0., 0., 0., 0. };
+    LHCOpticalFunctionsSet::Kinematics k_out;
+    rpod.optics->transport(k_in, k_out);
+    rpod.x0 = k_out.x;
+    rpod.y0 = k_out.y;
+
+    unique_ptr<TGraph> g_x_d_vs_xi = make_unique<TGraph>(ofs.getXiValues().size(), ofs.getXiValues().data(),
+      ofs.getFcnValues()[LHCOpticalFunctionsSet::exd].data());
+    ff->SetParameters(0., 0.);
+    g_x_d_vs_xi->Fit(ff.get(), "Q");
+    g_x_d_vs_xi->Fit(ff.get(), "Q");
+    g_x_d_vs_xi->Fit(ff.get(), "Q");
+    rpod.ch0 = ff->GetParameter(0) - rpod.x0;
+    rpod.ch1 = ff->GetParameter(1);
+
+    unique_ptr<TGraph> g_L_x_vs_xi = make_unique<TGraph>(ofs.getXiValues().size(), ofs.getXiValues().data(),
+      ofs.getFcnValues()[LHCOpticalFunctionsSet::eLx].data());
+    ff->SetParameters(0., 0.);
+    g_L_x_vs_xi->Fit(ff.get(), "Q");
+    g_L_x_vs_xi->Fit(ff.get(), "Q");
+    g_L_x_vs_xi->Fit(ff.get(), "Q");
```

Comment:
Add at least `N SERIAL` to the fit options (`0` may be useful as well).

Why is the `Fit()` called three times?